### PR TITLE
feat(relay): Wave 6 edge cases + battle test

### PR DESCRIPTION
**Pull request by seantokuzo**
## Summary
Wave 6 hardens the relay side of the sprite-agent wiring phase against every edge case the spec enumerates, and locks in regression coverage for the scenario tables.

### Persistence cascade (spec "Relay Persistence & Cascade Fallback")
- `SpriteMappingPersistence.load()` now distinguishes **missing** (ENOENT -> silent, normal new session) from **corrupt** (parse error / wrong schema / non-array mappings -> WARN log + null). Callers fall through to client-authoritative iOS re-send.
- `writeToDisk()` survives **ENOSPC**, **EIO**, and any other `fs` error — logs with an operator-friendly marker and returns `false`, never throws. In-memory mapping remains authoritative; iOS re-sends its mappings on reconnect and the next write retries.
- `delete()` differentiates ENOENT from real I/O errors.
- Constructor gains optional `{ baseDir, fs }` overrides for deterministic tests (no real filesystem dependency).

### Disconnect grace period (routes/ws.ts)
- Spec calls for a 30-min sprite-mapping grace period. The existing 10-min session timeout already satisfies this as a *stricter* upper bound — it also destroys the SDK session, which cascades to `fleetManager.destroySession(...) -> worker.btwQueue.dropForSession('Session ended')`. Documented the relationship and added a `SPRITE_GRACE_PERIOD_MAX_MS` log constant so the log line names the spec value.
- No behavior change in the happy path — the dropForSession reason is already present; this PR locks it into tests.

## Scenarios now covered by tests
`relay/src/sprites/__tests__/wave6-scenarios.test.ts`:

| Scenario | Spec ID | Test |
|---|---|---|
| Subagent crash mid-task | S1 | `sprite.unlink` with `reason: 'failed'`, /btw queue drains |
| Subagent completes with error | S2 | `sprite.unlink` with `reason: 'completed'`, queue drains |
| User dismisses subagent | S3 | `sprite.unlink` with `reason: 'dismissed'`, queue drains |
| Relay disconnect mid-subagent | S4 | Mapping preserved on disk + in memory, restored via load; queue survives until session-end |
| Fast spawn+complete (<1s) | S6 | link/unlink fire in order, metrics captured |
| All human sprites exhausted | S7 | Role-stable binding duplicates humans, **never** a dog CharacterType |
| Cold boot cleanup | S8 | listStale + delete reaps orphans, in-memory queue lost |
| New terminal session | S9 | No sprite state until first `agent.spawn` |
| Tap+send race (accept) | #9 | Mapping present at enqueue time -> accepted |
| Tap+send race (miss) | #9 | Mapping gone at enqueue -> error response |
| Disk-full end-to-end | | spawn + enqueue + unlink all work when every write throws ENOSPC |
| Corrupt file on reconnect | | `load()` null, iOS re-send repopulates via `saveImmediate` |
| Tool event cleanup on despawn | | `metricsStore.remove()` returns final snapshot for iOS to clear bubbles |

`relay/src/sprites/__tests__/sprite-mapping-persistence.test.ts`:
- Cascade tier 1: round-trip, ENOENT silent, delete, deleteAll.
- Cascade tier 2: invalid JSON, wrong schema, missing version, non-array mappings, injected EACCES.
- Disk-full survival: ENOSPC + EIO non-throw, transient-failure recovery.
- Cold-boot stale-file cleanup end-to-end.
- Path-traversal sanitization (silent no-op, safer than throwing).

## Test count
**Before:** 102 tests  
**After:** 139 tests (+37 Wave 6 tests across two new files)

## Non-goals / deferred
- **S5 desk overflow** — iOS-only sprite placement, not relay.
- **M2 cross-session banner** and **M3 bubble collision** — iOS UI.
- **Tool event cleanup wire signal** — iOS already clears tool bubbles on `agent.complete`/`agent.dismissed`; the metrics final snapshot is surfaced per Wave 5 contract. No new wire message was needed.

## Test plan
- [x] `cd relay && npm run build` clean
- [x] `cd relay && npm test` — 139/139 pass
- [x] `npx tsc --noEmit` clean
- [ ] Manual verification on device/iOS happens when Wave 6 lands in main

Co-Authored-By: Claude Opus 4.7 (1M context) <noreply@anthropic.com>

Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: relay/src/sprites/__tests__/wave6-scenarios.test.ts [hunk 1]
```diff
@@ -0,0 +1,696 @@
...
+    expect(entry).not.toBeNull();
+    expect(h.btwQueue.size).toBe(1);
+
+    // Unlink — still works, still emits drop.
+    h.unlinkAgent('sess-x', 'agent-x', 'completed');
+    expect(h.btwQueue.size).toBe(0);
+
+    // Force the debounce to flush any trailing saves.
+    await new Promise((r) => setTimeout(r, 2100));
+    persistence.dispose();
```
[the remark this answers is not in the document]
Comment (seantokuzo) [reply relay/src/sprites/__tests__/wave6-scenarios.test.ts]: Fixed in 061ba97 — swapped the `await new Promise((r) => setTimeout(r, 2100))` for `vi.useFakeTimers()` + `await vi.advanceTimersByTimeAsync(2100)` wrapped in a try/finally that restores real timers. The ENOSPC test now runs in ~3ms instead of 2100+ms. Full `npm test` stays at 1.31s with 139 tests passing.

File: relay/src/routes/ws.ts [hunk 2]
```diff
@@ -315,7 +341,11 @@ export function createWsRoute(deps: WsDeps): FastifyPluginAsync {
       const data = buildPersistedSession(sessionId);
       if (data) sessionPersistence.save(data);
       sessionManager.destroy(sessionId);
-      // Clean up sprite mappings for the timed-out session
+      // Clean up sprite mappings for the timed-out session.
+      // fleetManager.destroySession(...) already told the worker to run
+      // `btwQueue.dropForSession(sessionId, 'Session ended')` which emits
+      // 'dropped' for each pending /btw with the spec reason
+      // "Session ended (grace period expired)" semantics.
```
[the remark this answers is not in the document]
Comment (seantokuzo) [reply relay/src/routes/ws.ts]: Fixed in 061ba97 — comment now reads `"Session ended" (grace period expired — see SESSION_TIMEOUT_MS above)`, which matches the literal reason string passed into `dropForSession(sessionId, 'Session ended')` while still noting *why* the grace period triggered it.

File: relay/src/sprites/__tests__/sprite-mapping-persistence.test.ts [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+    return await readFile(path, 'utf-8');
+  } catch (err) {
+    if ((err as NodeJS.ErrnoException).code === 'ENOENT') return null;
+    throw err;
+  }
+}
+
+// Keep unused imports warning at bay — we keep `unlink` imported for future
+// use; marker so lint doesn't flip on it.
+void unlink;
```
[the remark this answers is not in the document]
Comment (seantokuzo) [reply relay/src/sprites/__tests__/sprite-mapping-persistence.test.ts]: Fixed in 061ba97 — removed both the unused `unlink` from the `node:fs/promises` import list and the `void unlink;` marker at the bottom of the file. Any future test that needs it can re-add the import locally.

File: relay/src/sprites/__tests__/sprite-mapping-persistence.test.ts [hunk 4]
```diff
@@ -0,0 +1,357 @@
...
+  });
+
+  afterEach(async () => {
+    persistence.dispose();
+    await rm(baseDir, { recursive: true, force: true });
+  });
+
+  // Path-traversal attempts silently no-op: filePath() throws internally,
+  // but the catch blocks in load() / delete() swallow it — safer than
+  // propagating an exception and leaking "which sessionIds are valid".
```
[the remark this answers is not in the document]
Comment (seantokuzo) [reply relay/src/sprites/__tests__/sprite-mapping-persistence.test.ts]: Fixed in 061ba97 — comment now states explicitly that load()/delete() funnel the thrown path-traversal error through their non-ENOENT WARN-log branch (so no throw propagates to the caller, but a WARN line IS emitted), and that only save()'s internal debounce-timer catch stays fully silent. The WARN is intentional: it helps operators spot suspect sessionIds without leaking valid-vs-invalid info back to the caller.
